dbms/src/Storages/DeltaMerge/tests/gtest_key_range.cpp

```diff
@@ -98,6 +98,56 @@ TEST(RowKeyRange_test, RedactRangeFromCommonHandle)
     Redact::setRedactLog(false); // restore flags
 }
 
+TEST(RowKey, ToNextKeyIntHandle)
+{
+    const auto key = RowKeyValue::fromHandle(20);
+    const auto next = key.toNext();
+    EXPECT_EQ("21", next.toDebugString());
+
+    {
+        const auto expected_next_int = RowKeyValue::fromHandle(21);
+        EXPECT_EQ(0, compare(next.toRowKeyValueRef(), expected_next_int.toRowKeyValueRef()));
+    }
+    {
+        const auto range_keys = std::make_shared<RegionRangeKeys>(
+            RecordKVFormat::genKey(1, 0),
+            RecordKVFormat::genKey(1, 21));
+        const auto range = RowKeyRange::fromRegionRange(
+            range_keys,
+            /* table_id */ 1,
+            /* is_common_handle */ false,
+            /* row_key_column_size */ 1);
+        EXPECT_EQ(0, compare(next.toRowKeyValueRef(), range.getEnd()));
+    }
+    // Note: The following does not work, because {20,00} will be regarded as Key=20 in RowKeyRange::fromRegionRange.
```

Comment:
It would be better if we could move the check before tiflash sync the data, but it seems hard.

- For pd-server, it knows nothing about "table" and "rowkey", so it can not check the split key is valid or not
- For tiflash-server, it does not (and should not?) do a schema check when applying Region split. So tiflash-server can only do this check when data is receive or query key range receive

* * *

Another approach to resolve this problem maybe to parse key {20, 00} as Key=21 in `RowKeyRange::fromRegionRange`. Can this avoid data loss when decoding snapshot data or handling queries with key range with '\00' suffix?

Reply:
> Another approach to resolve this problem maybe to parse key {20, 00} as Key=21 in RowKeyRange::fromRegionRange. Can this avoid data loss when decoding snapshot data or handling queries with key range with '\00' suffix?

I guess it could work.

My thinking is that, it would be nicer if we could enhance TiFlash's robustness to an extent of supporting **arbitrary** region ranges. `\00` is only one of the case. However this may be hard.

Another direction is to let TiFlash completely forbid arbitrary region ranges, but only accept correct ones (and always fail for incorrect ones). This could also be a good behavior if other components follow it.

It may not be a very good idea to stand in the middle of the two, as this may be not a not well-defined behavior now.

Reply:
yes, I agree